flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeGeneralization.java

```diff
@@ -0,0 +1,631 @@
+				return null;
+			}
+		}
+		return firstType;
+	}
+
+	private static @Nullable LogicalType findCommonTypePattern(LogicalType resultType, LogicalType type) {
+		if (hasFamily(resultType, DATETIME) && hasFamily(type, INTERVAL)) {
+			return resultType;
+		} else if (hasFamily(resultType, INTERVAL) && hasFamily(type, DATETIME)) {
+			return type;
+		} else if ((hasFamily(resultType, TIMESTAMP) || hasRoot(resultType, DATE)) && hasFamily(type, EXACT_NUMERIC)) {
+			return resultType;
+		} else if (hasFamily(resultType, EXACT_NUMERIC) && (hasFamily(type, TIMESTAMP) || hasRoot(type, DATE))) {
+			return type;
+		}
+		return null;
+	}
+
+	private static @Nullable LogicalType findCommonArrayType(List<LogicalType> normalizedTypes) {
+		final List<LogicalType> children = findCommonChildrenTypes(normalizedTypes);
+		if (children == null) {
+			return null;
+		}
+		return new ArrayType(children.get(0));
+	}
+
+	private static @Nullable LogicalType findCommonMultisetType(List<LogicalType> normalizedTypes) {
+		final List<LogicalType> children = findCommonChildrenTypes(normalizedTypes);
+		if (children == null) {
+			return null;
+		}
+		return new MultisetType(children.get(0));
+	}
+
+	private static @Nullable LogicalType findCommonRowType(List<LogicalType> normalizedTypes) {
+		final List<LogicalType> children = findCommonChildrenTypes(normalizedTypes);
+		if (children == null) {
+			return null;
+		}
+		final RowType firstType = (RowType) normalizedTypes.get(0);
+		final List<RowType.RowField> newFields = IntStream.range(0, children.size())
+			.mapToObj(pos -> {
+				final LogicalType newType = children.get(pos);
+				final RowType.RowField originalField = firstType.getFields().get(pos);
+				if (originalField.getDescription().isPresent()) {
+					return new RowType.RowField(
+						originalField.getName(),
+						newType,
+						originalField.getDescription().get());
+				} else {
+					return new RowType.RowField(
+						originalField.getName(),
+						newType);
+				}
+			})
+			.collect(Collectors.toList());
+		return new RowType(newFields);
+	}
+
+	private static @Nullable List<LogicalType> findCommonChildrenTypes(List<LogicalType> normalizedTypes) {
+		final LogicalType firstType = normalizedTypes.get(0);
+		final LogicalTypeRoot typeRoot = firstType.getTypeRoot();
+		final int numberOfChildren = firstType.getChildren().size();
+
+		for (LogicalType type : normalizedTypes) {
+			// all types must have the same root
+			if (type.getTypeRoot() != typeRoot) {
+				return null;
+			}
+			// all types must have the same number of children
+			if (type.getChildren().size() != numberOfChildren) {
+				return null;
+			}
+		}
+
+		// recursively compute column-wise least restrictive
+		final List<LogicalType> resultChildren = new ArrayList<>(numberOfChildren);
+		for (int i = 0; i < numberOfChildren; i++) {
+			final int childPos = i;
+			final Optional<LogicalType> childType = findCommonType(new AbstractList<LogicalType>() {
```

Comment:
I would extract this anonymous class to give it a better name and maybe a bit of explanation. 

I guess we could name it `ChildTypeView` or sth similar `*View`?